### PR TITLE
Fallback behavior to ensure in-proc payload compatibility with `dotnet-isolated` as the `FUNCTIONS_WORKER_RUNTIME` value

### PR DESCRIPTION
**Pull request by kshyju**
This update addresses the issue of running a function app where the `FUNCTIONS_WORKER_RUNTIME` setting does not match the function metadata from the deployed code. For example, if a function app is configured with `"dotnet-isolated"` as the `FUNCTIONS_WORKER_RUNTIME` but has an in-process (in-proc) app payload deployed, this scenario previously worked in version 4.34.

However, in version 4.35, a change was introduced (see [this PR](https://github.com/Azure/azure-functions-dotnet-worker/pull/2552)) that restricts the use of the dotnet-isolated worker configuration to only when the app is running in placeholder mode. This change was intentional to ensure that `FunctionsNetHost` starts only in placeholder mode. However, it inadvertently caused issues for incorrectly configured apps (i.e., those with an in-proc payload and a `"dotnet-isolated"` `FUNCTIONS_WORKER_RUNTIME` value), as this new behavior prevents the creation of the worker configuration instance, thereby stopping the function app from running.

This PR includes changes to temporarily support this use case while also emitting a diagnostic warning to customers. The intention is to remove this support in a future release.

#### Why it works in 4.34

The reason it works in 4.34 is we update the "Language" property of function metadata to "dotnet-isolated" (the language present in the workerconfig instance). Code [here](https://github.com/Azure/azure-functions-host/blob/ff0dea77c3de615a9a19f000595dc56262f47ad4/src/WebJobs.Script/Host/HostFunctionMetadataProvider.cs#L166) and [here](https://github.com/Azure/azure-functions-host/blob/ff0dea77c3de615a9a19f000595dc56262f47ad4/src/WebJobs.Script/Host/HostFunctionMetadataProvider.cs#L200-L204)). Having this change (Language in function metadata entries matching with FUNCTIONS_WORKER_RUNTIME
) and having at least one workerconfig instance to match the worker runtime value (dotnet-isolated) make things works. 
### Pull request checklist

**IMPORTANT**: Currently, changes must be backported to the `in-proc` branch to be included in Core Tools and non-Flex deployments.

* [ ] Backporting to the `in-proc` branch is not required
    * Otherwise: Link to backporting PR - https://github.com/Azure/azure-functions-host/pull/10449
* [x] My changes **do not** require documentation changes
    * [ ] Otherwise: Documentation issue linked to PR
* [ ] My changes **should not** be added to the release notes for the next release
    * [x] Otherwise: I've added my notes to `release_notes.md`
* [ ] My changes **do not** need to be backported to a previous version
    * [x ] Otherwise: Backport tracked by issue/PR #issue_or_pr - will follow
* [ ] My changes **do not** require diagnostic events changes
    * Otherwise: Added diag event. will work on docs change.
* [x] I have added all required tests (Unit tests, E2E tests)

### Additional information

Additional PR information

### CODE REVIEW

File: src/WebJobs.Script/Diagnostics/DiagnosticEventConstants.cs [hunk 1]
```diff
@@ -31,5 +31,8 @@ internal static class DiagnosticEventConstants
 
         public const string NonHISSecretLoaded = "AZFD0012";
         public const string NonHISSecretLoadedHelpLink = "https://aka.ms/functions-non-his-secrets";
+
+        public const string WorkerRuntimeDoesNotMatchWithFunctionMetadataErrorCode = "AZFD0013";
+        public const string WorkerRuntimeDoesNotMatchWithFunctionMetadataHelpLink = "https://aka.ms/functions-invalid-worker-runtime";
```
Comment (kshyju) [inline src/WebJobs.Script/Diagnostics/DiagnosticEventConstants.cs:36]: yet to create this.

File: test/WebJobs.Script.Tests.Integration/WebHostEndToEnd/WebJobsStartupEndToEndTests.cs [hunk 2]
```diff
@@ -45,6 +46,27 @@ public async Task ExternalStartup_Succeeds()
             }
         }
 
+        [Fact]
+        public async Task InProcAppsWorkWithDotnetIsolatedAsFunctionWorkerRuntimeValue()
```
Comment (kshyju) [inline test/WebJobs.Script.Tests.Integration/WebHostEndToEnd/WebJobsStartupEndToEndTests.cs:50]: This test fails in 4.35

File: src/WebJobs.Script/Host/ScriptHost.cs [hunk 3]
```diff
@@ -773,11 +773,39 @@ private void TrySetDirectType(FunctionMetadata metadata)
             }
         }
 
+        // Ensure customer deployed application payload matches with the worker runtime configured for the function app and log a warning if not.
+        // If a customer has "dotnet-isolated" worker runtime configured for the function app, and then they deploy an in-proc app payload, this will warn/error
+        // If there is a mismatch, the method will return false, else true.
+        internal static bool ValidateAndLogRuntimeMismatchForDotnetIsolated(IEnumerable<FunctionMetadata> functionMetadata, string workerRuntime, IOptions<FunctionsHostingConfigOptions> hostingConfigOptions, ILogger logger)
+        {
+            if (functionMetadata.Any() && string.Equals(workerRuntime, RpcWorkerConstants.DotNetIsolatedLanguageWorkerName, StringComparison.OrdinalIgnoreCase) && !Utility.ContainsAnyFunctionMatchingWorkerRuntime(functionMetadata, workerRuntime))
```
Comment (kshyju) [inline src/WebJobs.Script/Host/ScriptHost.cs:781]: I purposefully used very specific conditions (checking for dotnet-isolated) to limit the impact of the change.

File: src/WebJobs.Script/Utility.cs [hunk 4]
```diff
@@ -748,10 +744,20 @@ private static bool ContainsFunctionWithWorkerRuntime(IEnumerable<FunctionMetada
             {
                 return functions.Any(f => dotNetLanguages.Any(l => l.Equals(f.Language, StringComparison.OrdinalIgnoreCase)));
             }
+
+            return ContainsAnyFunctionMatchingWorkerRuntime(functions, workerRuntime);
+        }
+
+        /// <summary>
+        /// Inspect the functions metadata to determine if atleast one function is of the specified worker runtime.
```
Comment (satvu) [inline src/WebJobs.Script/Utility.cs]:
```suggestion
        /// Inspect the functions metadata to determine if at least one function is of the specified worker runtime.
```

File: src/WebJobs.Script/Host/ScriptHost.cs [hunk 5]
```diff
@@ -773,11 +773,39 @@ private void TrySetDirectType(FunctionMetadata metadata)
...
+            {
+                string baseMessage = $"The '{EnvironmentSettingNames.FunctionWorkerRuntime}' is not matching with worker runtime of function metadata of deployed function app artifacts. See {DiagnosticEventConstants.WorkerRuntimeDoesNotMatchWithFunctionMetadataHelpLink} for more information.";
+
+                if (hostingConfigOptions.Value.ThrowOnFunctionsWorkerRuntimeMismatchWithMetadataFromPayload)
+                {
+                    logger.LogDiagnosticEventError(DiagnosticEventConstants.WorkerRuntimeDoesNotMatchWithFunctionMetadataErrorCode, baseMessage, DiagnosticEventConstants.WorkerRuntimeDoesNotMatchWithFunctionMetadataHelpLink, null);
```
Comment (mattchenderson) [inline src/WebJobs.Script/Host/ScriptHost.cs:788]: Remind me, will this code path also ensure that this shows up in customer app logs as well? Like App Insights?

File: src/WebJobs.Script/Host/ScriptHost.cs [hunk 6]
```diff
@@ -773,11 +773,39 @@ private void TrySetDirectType(FunctionMetadata metadata)
...
         }
 
+        // Ensure customer deployed application payload matches with the worker runtime configured for the function app and log a warning if not.
+        // If a customer has "dotnet-isolated" worker runtime configured for the function app, and then they deploy an in-proc app payload, this will warn/error
+        // If there is a mismatch, the method will return false, else true.
+        internal static bool ValidateAndLogRuntimeMismatchForDotnetIsolated(IEnumerable<FunctionMetadata> functionMetadata, string workerRuntime, IOptions<FunctionsHostingConfigOptions> hostingConfigOptions, ILogger logger)
+        {
+            if (functionMetadata.Any() && string.Equals(workerRuntime, RpcWorkerConstants.DotNetIsolatedLanguageWorkerName, StringComparison.OrdinalIgnoreCase) && !Utility.ContainsAnyFunctionMatchingWorkerRuntime(functionMetadata, workerRuntime))
+            {
+                string baseMessage = $"The '{EnvironmentSettingNames.FunctionWorkerRuntime}' is not matching with worker runtime of function metadata of deployed function app artifacts. See {DiagnosticEventConstants.WorkerRuntimeDoesNotMatchWithFunctionMetadataHelpLink} for more information.";
```
Comment (mattchenderson) [inline src/WebJobs.Script/Host/ScriptHost.cs]: Should the format string for this message also just be a constant?

Regardless, small wording change suggested:

> $"The '{EnvironmentSettingNames.FunctionWorkerRuntime}' is set to '{}', which does not match the worker runtime metadata found in the deployed function app artifacts. The deployed artifacts are for worker runtime '{}'. See {DiagnosticEventConstants.WorkerRuntimeDoesNotMatchWithFunctionMetadataHelpLink} for more information."

With the appropriate extra interpolation of the actual values involved.

File: src/WebJobs.Script/Host/ScriptHost.cs [hunk 7]
```diff
@@ -773,11 +773,39 @@ private void TrySetDirectType(FunctionMetadata metadata)
...
+
+                if (hostingConfigOptions.Value.ThrowOnFunctionsWorkerRuntimeMismatchWithMetadataFromPayload)
+                {
+                    logger.LogDiagnosticEventError(DiagnosticEventConstants.WorkerRuntimeDoesNotMatchWithFunctionMetadataErrorCode, baseMessage, DiagnosticEventConstants.WorkerRuntimeDoesNotMatchWithFunctionMetadataHelpLink, null);
+                    throw new HostInitializationException(baseMessage);
+                }
+
+                string warningMessage = baseMessage + " The application will continue to run, but may throw an exception in a future release.";
```
Comment (mattchenderson) [inline src/WebJobs.Script/Host/ScriptHost.cs]: Should this just say "in the future" here? The release itself is not controlled by the user. And the Feature flag doesn't itself require a "release" even in our standard usage.

File: src/WebJobs.Script/Host/ScriptHost.cs [hunk 8]
```diff
@@ -819,6 +847,17 @@ internal async Task<Collection<FunctionDescriptor>> GetFunctionDescriptorsAsync(
             return functionDescriptors;
         }
 
+        private static void UpdateFunctionMetadataLanguageForDotnetAssembly(IEnumerable<FunctionMetadata> functions, string workerRuntime)
+        {
+            foreach (var function in functions)
+            {
+                if (function.Language == DotNetScriptTypes.DotNetAssembly)
+                {
+                    function.Language = workerRuntime;
```
Comment (satvu) [inline src/WebJobs.Script/Host/ScriptHost.cs:864]: Changing the function metadata language to match is enough to have customer apps running? In this misconfigured scenario does the app end up running like an inproc or isolated payload?

Comment (kshyju) [reply src/WebJobs.Script/Host/ScriptHost.cs:864]: Yes. The reason it works in 4.34 is we update the "Language" property of function metadata to "dotnet-isolated" (the language present in the workerconfig instance). Code [here](https://github.com/Azure/azure-functions-host/blob/ff0dea77c3de615a9a19f000595dc56262f47ad4/src/WebJobs.Script/Host/HostFunctionMetadataProvider.cs#L166) and [here](https://github.com/Azure/azure-functions-host/blob/ff0dea77c3de615a9a19f000595dc56262f47ad4/src/WebJobs.Script/Host/HostFunctionMetadataProvider.cs#L200-L204)). Having this change (Language in function metadata entries matching with FUNCTIONS_WORKER_RUNTIME
) and having at least one workerconfig instance to match the worker runtime value (dotnet-isolated) make things works.

File: src/WebJobs.Script/Host/ScriptHost.cs [hunk 9]
```diff
@@ -773,11 +773,39 @@ private void TrySetDirectType(FunctionMetadata metadata)
             }
         }
 
+        // Ensure customer deployed application payload matches with the worker runtime configured for the function app and log a warning if not.
+        // If a customer has "dotnet-isolated" worker runtime configured for the function app, and then they deploy an in-proc app payload, this will warn/error
+        // If there is a mismatch, the method will return false, else true.
+        internal static bool ValidateAndLogRuntimeMismatchForDotnetIsolated(IEnumerable<FunctionMetadata> functionMetadata, string workerRuntime, IOptions<FunctionsHostingConfigOptions> hostingConfigOptions, ILogger logger)
+        {
```
Comment (brettsam) [inline src/WebJobs.Script/Host/ScriptHost.cs:780]: null-check functionMetadata?

File: src/WebJobs.Script/Workers/Rpc/FunctionRegistration/RpcFunctionInvocationDispatcher.cs [hunk 10]
```diff
@@ -308,13 +311,19 @@ public async Task InitializeAsync(IEnumerable<FunctionMetadata> functions, Cance
             }
             else
             {
-                _processStartupInterval = workerConfig.CountOptions.ProcessStartupInterval;
-                _restartWait = workerConfig.CountOptions.ProcessRestartInterval;
-                _shutdownTimeout = workerConfig.CountOptions.ProcessShutdownTimeout;
+                _processStartupInterval = workerConfig?.CountOptions?.ProcessStartupInterval ?? _defaultProcessStartupInterval;
```
Comment (brettsam) [inline src/WebJobs.Script/Workers/Rpc/FunctionRegistration/RpcFunctionInvocationDispatcher.cs:314]: are these changes related?

File: src/WebJobs.Script/Workers/Rpc/FunctionRegistration/RpcFunctionInvocationDispatcher.cs [hunk 11]
```diff
@@ -308,13 +311,19 @@ public async Task InitializeAsync(IEnumerable<FunctionMetadata> functions, Cance
...
+                _shutdownTimeout = workerConfig?.CountOptions.ProcessShutdownTimeout ?? _defaultProcessShutdownInterval;
             }
             ErrorEventsThreshold = 3 * await _maxProcessCount.Value;
 
-            if (Utility.IsSupportedRuntime(_workerRuntime, _workerConfigs) || _environment.IsMultiLanguageRuntimeEnvironment())
+            // If the configured worker runtime is "dotnet-isolated" and no worker config is found, we should assume that this was caused by
+            //  1. App did not start in placeholder mode (so that the dotnet-isolated worker config was not initialized because of https://github.com/Azure/azure-functions-dotnet-worker/pull/2552)
+            //  2. App payload deployed is not "dotnet-isolated" (but "in-proc")
+            // In this case, we want to go ahead and initialize a language worker channel for the runtime.
+            var missingWorkerConfigForDotnetIsolated = !_workerConfigs.Any() && _workerRuntime == RpcWorkerConstants.DotNetIsolatedLanguageWorkerName;
```
Comment (brettsam) [inline src/WebJobs.Script/Workers/Rpc/FunctionRegistration/RpcFunctionInvocationDispatcher.cs]: we can start a channel without a worker config? does it just not launch a process?

File: release_notes.md [hunk 12]
```diff
@@ -18,5 +18,6 @@
 - Worker termination path updated with sanitized logging (#10367)
 - Avoid redundant DiagnosticEvents error message (#10395)
 - Added logic to shim older versions of the .NET Worker JsonFunctionProvider to ensure backwards compatibility (#10410)
+- Added fallback behavior when FUNCTIONS_WORKER_RUNTIME does not match with metadata from deployed app payload.
```
Comment (mathewc) [inline release_notes.md]: nit: I'd clarify here that this fallback behavior only applies to .NET Isolated

File: src/WebJobs.Script/Config/FunctionsHostingConfigOptions.cs [hunk 13]
```diff
@@ -183,6 +183,14 @@ internal bool ThrowOnMissingFunctionsWorkerRuntime
             }
         }
 
+        internal bool ThrowOnFunctionsWorkerRuntimeMismatchWithMetadataFromPayload
+        {
+            get
+            {
+                return GetFeature(RpcWorkerConstants.ThrowOnFunctionsWorkerRuntimeMismatchWithMetadataFromPayload) == "1";
```
Comment (mathewc) [inline src/WebJobs.Script/Config/FunctionsHostingConfigOptions.cs]: Perhaps use the GetFeatureAsBooleanOrDefault helper which seems to be what many of the other boolean configs are using?

File: src/WebJobs.Script/Host/ScriptHost.cs (hunk 7 above)
Comment (mathewc) [reply src/WebJobs.Script/Host/ScriptHost.cs]: Agreed. I think this should say "The application will be allowed to run now, but in the future this may become a startup failure."

File: src/WebJobs.Script/Config/FunctionsHostingConfigOptions.cs (hunk 13 above)
Comment (mathewc) [reply src/WebJobs.Script/Config/FunctionsHostingConfigOptions.cs]: Also, perhaps we name this config WorkerRuntimeStrictValidationEnabled

File: src/WebJobs.Script/Utility.cs [hunk 14]
```diff
@@ -644,10 +644,6 @@ internal static void VerifyFunctionsMatchSpecifiedLanguage(IEnumerable<FunctionM
                 {
                     throw new HostInitializationException($"Found functions with more than one language. Select a language for your function app by specifying {RpcWorkerConstants.FunctionWorkerRuntimeSettingName} AppSetting");
                 }
-                else
```
Comment (mathewc) [inline src/WebJobs.Script/Utility.cs:647]: Why are you deleting this? Isn't this .NET case you're looking at fall into the bucket of "single language"?

File: src/WebJobs.Script/Workers/Rpc/FunctionRegistration/RpcFunctionInvocationDispatcher.cs (hunk 10 above)
Comment (kshyju) [reply src/WebJobs.Script/Workers/Rpc/FunctionRegistration/RpcFunctionInvocationDispatcher.cs:314]: yes, since there will not be a `workerConfig`, this code throws a NRE. (we did not get to this state in 4.35 because we threw `new HostInitializationException($"Did not find functions with language [{workerRuntime}].");`)

File: src/WebJobs.Script/Utility.cs (hunk 14 above)
Comment (kshyju) [reply src/WebJobs.Script/Utility.cs:647]: yes. that is needed. I added it back in the last iteration, but now it conditionally throws based on the `throwOnMismatch` parameter value. With the new changes, we do not want to throw in the inproc payload + DNI FWRuntime value.

File: src/WebJobs.Script/Host/ScriptHost.cs (hunk 9 above)
Comment (kshyju) [reply src/WebJobs.Script/Host/ScriptHost.cs:780]: Added null check in the latest iteration.

File: release_notes.md (hunk 12 above)
Comment (kshyju) [reply release_notes.md]: Updated to "Added fallback behavior to ensure in-proc payload compatibility with "dotnet-isolated" as the `FUNCTIONS_WORKER_RUNTIME` value"

File: src/WebJobs.Script/Workers/Rpc/FunctionRegistration/RpcFunctionInvocationDispatcher.cs (hunk 11 above)
Comment (mathewc) [inline src/WebJobs.Script/Workers/Rpc/FunctionRegistration/RpcFunctionInvocationDispatcher.cs]: We should do a case insensitive match on _workerRuntime, since it's a user supplied value, like we do elsewhere

File: src/WebJobs.Script/Host/ScriptHost.cs [hunk 15]
```diff
@@ -773,14 +773,49 @@ private void TrySetDirectType(FunctionMetadata metadata)
...
+        }
+
         internal async Task<Collection<FunctionDescriptor>> GetFunctionDescriptorsAsync(IEnumerable<FunctionMetadata> functions, IEnumerable<FunctionDescriptorProvider> descriptorProviders, string workerRuntime, CancellationToken cancellationToken)
         {
             Collection<FunctionDescriptor> functionDescriptors = new Collection<FunctionDescriptor>();
             if (!cancellationToken.IsCancellationRequested)
             {
+                bool throwOnWorkerRuntimeAndPayloadMetadataMismatch = true;
+                // this dotnet isolated specific logic is temporary to ensure in-proc payload compatibility with "dotnet-isolated" as the FUNCTIONS_WORKER_RUNTIME value.
+                if (string.Equals(workerRuntime, RpcWorkerConstants.DotNetIsolatedLanguageWorkerName, StringComparison.OrdinalIgnoreCase))
```
Comment (kshyju) [inline src/WebJobs.Script/Host/ScriptHost.cs:807]: I purposefully used very specific conditions (checking for dotnet-isolated) to limit the impact of the change to keep it to the problem we are fixing (in-proc app should work with dotnet isolated as FWR)

File: src/WebJobs.Script/Host/ScriptHost.cs (hunk 5 above)
Comment (kshyju) [reply src/WebJobs.Script/Host/ScriptHost.cs:788]: I am not 100% sure about that. @soninaren Do you know?

File: src/WebJobs.Script/Workers/Rpc/FunctionRegistration/RpcFunctionInvocationDispatcher.cs (hunk 11 above)
Comment (kshyju) [reply src/WebJobs.Script/Workers/Rpc/FunctionRegistration/RpcFunctionInvocationDispatcher.cs]: Good point. This change is not even required. I removed it in the latest iteration. We do not start a language worker channel.

File: src/WebJobs.Script/Host/ScriptHost.cs (hunk 5 above)
Comment (cjaliaga) [reply src/WebJobs.Script/Host/ScriptHost.cs:788]: They should appear, I just ran a test to confirm with latest commit in the branch.

image

**Review by cjaliaga — APPROVED** [review]

File: src/WebJobs.Script/Host/ScriptHost.cs [hunk 16]
```diff
@@ -773,14 +773,49 @@ private void TrySetDirectType(FunctionMetadata metadata)
...
+
+            return true;
+        }
+
         internal async Task<Collection<FunctionDescriptor>> GetFunctionDescriptorsAsync(IEnumerable<FunctionMetadata> functions, IEnumerable<FunctionDescriptorProvider> descriptorProviders, string workerRuntime, CancellationToken cancellationToken)
         {
             Collection<FunctionDescriptor> functionDescriptors = new Collection<FunctionDescriptor>();
             if (!cancellationToken.IsCancellationRequested)
             {
+                bool throwOnWorkerRuntimeAndPayloadMetadataMismatch = true;
```
Comment (mathewc) [inline src/WebJobs.Script/Host/ScriptHost.cs:805]: What's the difference between this and ThrowOnFunctionsWorkerRuntimeMismatchWithMetadataFromPayload? In once case above you're using the hosting config value as expected, but here you're not?

File: src/WebJobs.Script/Host/ScriptHost.cs (hunk 6 above)
Comment (mattchenderson) [reply src/WebJobs.Script/Host/ScriptHost.cs]: The latest iteration is closer:

`string baseMessage = $"The '{EnvironmentSettingNames.FunctionWorkerRuntime}' is set to '{workerRuntime}', which does not match the worker runtime metadata found in the deployed function app artifacts. See {DiagnosticEventConstants.WorkerRuntimeDoesNotMatchWithFunctionMetadataHelpLink} for more information.";`

I would still like to see the inclusion of what it found ("The deployed artifacts are for worker runtime '{}'." in above suggestion).

Comment (kshyju) [reply src/WebJobs.Script/Host/ScriptHost.cs]: Updated in the latest iteration. See the test for an example message.

File: src/WebJobs.Script/Config/FunctionsHostingConfigOptions.cs (hunk 13 above)
Comment (kshyju) [reply src/WebJobs.Script/Config/FunctionsHostingConfigOptions.cs]: Renamed to `WorkerRuntimeStrictValidationEnabled` as suggested.

File: src/WebJobs.Script/Host/ScriptHost.cs (hunk 6 above)
Comment (mattchenderson) [reply src/WebJobs.Script/Host/ScriptHost.cs]: Awesome - thanks!

**Review by mathewc — APPROVED** [review]

File: src/WebJobs.Script/Workers/Rpc/FunctionRegistration/RpcFunctionInvocationDispatcher.cs (hunk 11 above)
Comment (kshyju) [reply src/WebJobs.Script/Workers/Rpc/FunctionRegistration/RpcFunctionInvocationDispatcher.cs]: This change (string comparison for which you made the comment) was removed in a later iteration as it is not needed.

**Review by brettsam — APPROVED** [review]
